File: student_result_system.py
# Features:
# Add student records
# Store marks of subjects
# Calculate total marks
# Calculate percentage
# Assign grades
# Display student results
# Search student results
# Python Code:
# Student Result System

students = {}

def calculate_grade(percentage):
    if percentage >= 90:
        return "A+"
    elif percentage >= 80:
        return "A"
    elif percentage >= 70:
        return "B"
    elif percentage >= 60:
        return "C"
    elif percentage >= 50:
        return "D"
    else:
        return "F"


def add_student():
    roll_no = input("Enter Roll Number: ")
    name = input("Enter Student Name: ")

    marks = {}
    subjects = ["Math", "Physics", "Computer"]

    for subject in subjects:
        marks[subject] = int(input(f"Enter {subject} marks: "))

    total = sum(marks.values())
    percentage = total / len(subjects)

    grade = calculate_grade(percentage)

    students[roll_no] = {
        "name": name,
        "marks": marks,
        "total": total,
        "percentage": percentage,
        "grade": grade
    }

    print("Student added successfully!\n")


def display_results():
    if not students:
        print("No records found")
        return

    for roll, data in students.items():
        print("\n----------------------")
        print("Roll Number:", roll)
        print("Name:", data["name"])
        print("Marks:", data["marks"])
        print("Total:", data["total"])
        print("Percentage:", data["percentage"], "%")
        print("Grade:", data["grade"])


def search_student():
    roll = input("Enter Roll Number to search: ")

    if roll in students:
        data = students[roll]

        print("\nStudent Result")
        print("Name:", data["name"])
        print("Marks:", data["marks"])
        print("Percentage:", data["percentage"], "%")
        print("Grade:", data["grade"])
    else:
        print("Student not found")


while True:
    print("\n===== Student Result System =====")
    print("1. Add Student")
    print("2. Display Results")
    print("3. Search Student")
    print("4. Exit")

    choice = input("Enter choice: ")

    if choice == "1":
        add_student()

    elif choice == "2":
        display_results()

    elif choice == "3":
        search_student()

    elif choice == "4":
        print("Program Ended")
        break

    else:
        print("Invalid choice")

# Example Output:
# ===== Student Result System =====
# 1. Add Student
# 2. Display Results
# 3. Search Student
# 4. Exit

# Enter choice: 1

# Enter Roll Number: 101
# Enter Student Name: Fatima
# Enter Math marks: 90
# Enter Physics marks: 85
# Enter Computer marks: 95

# Student added successfully!
# Concepts Practiced:

# ✅ Variables
# ✅ Input/Output
# ✅ Functions
# ✅ Dictionaries
# ✅ Loops
# ✅ Conditional statements
# ✅ CRUD operations

# Next improvements you can add:

# Save records permanently using a .txt or .json file
# Add GUI using Tkinter
# Add login system for teachers/admin
# Generate result cards as PDF
# Use SQLite database

File: student_result_system.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class StudentResultSystem {

    static Scanner sc = new Scanner(System.in);
    static HashMap<String, Student> students = new HashMap<>();

    static class Student {
        String name;
        HashMap<String, Integer> marks;
        int total;
        double percentage;
        String grade;

        Student(String name, HashMap<String, Integer> marks, int total, double percentage, String grade) {
            this.name = name;
            this.marks = marks;
            this.total = total;
            this.percentage = percentage;
            this.grade = grade;
        }
    }

    public static String calculateGrade(double percentage) {
        if (percentage >= 90) {
            return "A+";
        } else if (percentage >= 80) {
            return "A";
        } else if (percentage >= 70) {
            return "B";
        } else if (percentage >= 60) {
            return "C";
        } else if (percentage >= 50) {
            return "D";
        } else {
            return "F";
        }
    }

    public static void addStudent() {
        System.out.print("Enter Roll Number: ");
        String rollNo = sc.nextLine();

        System.out.print("Enter Student Name: ");
        String name = sc.nextLine();

        HashMap<String, Integer> marks = new HashMap<>();

        String[] subjects = {"Math", "Physics", "Computer"};

        int total = 0;

        for (String subject : subjects) {
            System.out.print("Enter " + subject + " marks: ");
            int mark = sc.nextInt();
            sc.nextLine();
            marks.put(subject, mark);
            total += mark;
        }

        double percentage = (double) total / subjects.length;
        String grade = calculateGrade(percentage);

        Student student = new Student(name, marks, total, percentage, grade);
        students.put(rollNo, student);

        System.out.println("Student added successfully!\n");
    }

    public static void displayResults() {
        if (students.isEmpty()) {
            System.out.println("No records found");
            return;
        }

        for (Map.Entry<String, Student> entry : students.entrySet()) {
            String roll = entry.getKey();
            Student data = entry.getValue();

            System.out.println("\n----------------------");
            System.out.println("Roll Number: " + roll);
            System.out.println("Name: " + data.name);
            System.out.println("Marks: " + data.marks);
            System.out.println("Total: " + data.total);
            System.out.printf("Percentage: %.2f%%\n", data.percentage);
            System.out.println("Grade: " + data.grade);
        }
    }

    public static void searchStudent() {
        System.out.print("Enter Roll Number to search: ");
        String roll = sc.nextLine();

        if (students.containsKey(roll)) {
            Student data = students.get(roll);

            System.out.println("\nStudent Result");
            System.out.println("Name: " + data.name);
            System.out.println("Marks: " + data.marks);
            System.out.printf("Percentage: %.2f%%\n", data.percentage);
            System.out.println("Grade: " + data.grade);
        } else {
            System.out.println("Student not found");
        }
    }

    public static void main(String[] args) {

        while (true) {
            System.out.println("\n===== Student Result System =====");
            System.out.println("1. Add Student");
            System.out.println("2. Display Results");
            System.out.println("3. Search Student");
            System.out.println("4. Exit");

            System.out.print("Enter choice: ");
            String choice = sc.nextLine();

            switch (choice) {
                case "1":
                    addStudent();
                    break;

                case "2":
                    displayResults();
                    break;

                case "3":
                    searchStudent();
                    break;

                case "4":
                    System.out.println("Program Ended");
                    return;

                default:
                    System.out.println("Invalid choice");
            }
        }
    }
}
